homeassistant/components/logi_circle/camera.py

```diff
@@ -1,113 +1,57 @@
-})
-
-LOGI_CIRCLE_SERVICE_RECORD = CAMERA_SERVICE_SCHEMA.extend({
-    vol.Required(ATTR_FILENAME): cv.template,
-    vol.Required(ATTR_DURATION): cv.positive_int
-})
-
 
 async def async_setup_platform(
         hass, config, async_add_entities, discovery_info=None):
-    """Set up a Logi Circle Camera."""
-    devices = hass.data[LOGI_CIRCLE_DOMAIN]
+    """Set up a Logi Circle Camera. Obsolete."""
+    _LOGGER.warning(
+        "Logi Circle no longer works with camera platform configuration")
 
-    cameras = []
-    for device in devices:
-        cameras.append(LogiCam(device, config))
 
-    async_add_entities(cameras, True)
+async def async_setup_entry(hass, entry, async_add_entities):
+    """Set up a Logi Circle Camera based on a config entry."""
+    devices = await hass.data[LOGI_CIRCLE_DOMAIN][ATTR_API].cameras
+    ffmpeg = hass.data[DATA_FFMPEG]
 
-    async def service_handler(service):
-        """Dispatch service calls to target entities."""
-        params = {key: value for key, value in service.data.items()
-                  if key != ATTR_ENTITY_ID}
-        entity_ids = service.data.get(ATTR_ENTITY_ID)
-        if entity_ids:
-            target_devices = [dev for dev in cameras
-                              if dev.entity_id in entity_ids]
-        else:
-            target_devices = cameras
-
-        for target_device in target_devices:
-            if service.service == SERVICE_SET_CONFIG:
-                await target_device.set_config(**params)
-            if service.service == SERVICE_LIVESTREAM_SNAPSHOT:
-                await target_device.livestream_snapshot(**params)
-            if service.service == SERVICE_LIVESTREAM_RECORD:
-                await target_device.download_livestream(**params)
-
-    hass.services.async_register(
-        DOMAIN, SERVICE_SET_CONFIG, service_handler,
-        schema=LOGI_CIRCLE_SERVICE_SET_CONFIG)
-
-    hass.services.async_register(
-        DOMAIN, SERVICE_LIVESTREAM_SNAPSHOT, service_handler,
-        schema=LOGI_CIRCLE_SERVICE_SNAPSHOT)
-
-    hass.services.async_register(
-        DOMAIN, SERVICE_LIVESTREAM_RECORD, service_handler,
-        schema=LOGI_CIRCLE_SERVICE_RECORD)
+    cameras = [LogiCam(device, entry, ffmpeg)
+               for device in devices]
+
+    async_add_entities(cameras, True)
 
 
 class LogiCam(Camera):
     """An implementation of a Logi Circle camera."""
 
-    def __init__(self, camera, device_info):
+    def __init__(self, camera, device_info, ffmpeg):
         """Initialize Logi Circle camera."""
         super().__init__()
         self._camera = camera
         self._name = self._camera.name
         self._id = self._camera.mac_address
         self._has_battery = self._camera.supports_feature('battery_level')
+        self._ffmpeg = ffmpeg
+
+    async def async_added_to_hass(self):
+        """Make entity globally accessible for use in service handler."""
+        self.hass.data[LOGI_CIRCLE_DOMAIN]['entities']['camera'].append(self)
```

Comment:
We're not allowed to store entities in a shared container. Entities are a platform concern.

Connect the entity methods to the signals of the servics for the dispatch helper here instead.

Reply:
I've added `SIGNAL_LOGI_CIRCLE_RECONFIGURE`, `SIGNAL_LOGI_CIRCLE_RECORD` and `SIGNAL_LOGI_CIRCLE_SNAPSHOT` for the 3 logi circle services and added listeners to the camera entity. Hopefully that's the right approach. 

Addressed in commit [3aaedb86](https://github.com/home-assistant/home-assistant/pull/20624/commits/3aaedb86d735509f2ab939668741aeb0ce10b973).